agents/magpie/magpie_agent.py

```diff
@@ -0,0 +1,743 @@
+            y += offset[1]
+
+            name = f"{stream_id}/det_{i}"
+            eqs = [f'{name}/raw', f'* {name}/rms rms_scale']
+            value_names = [f'{name}/raw', f'{name}/rms']
+            eq_labels = ['raw', 'rms']
+
+            fp.add_vis_elem(name, x, y, rot, value_names, eqs, eq_labels,
+                            cmaps[cidx], template, i, [True, False])
+
+        return fp
+
+
+class MagpieAgent:
+    """
+    Agent for processing streamed G3Frames, and sending data to lyrebird.
+
+    Attributes
+    -----------
+    target_rate : float
+        This is the target sample rate of data to be sent to lyrebird.
+        Incoming data will be downsampled to this rate before being sent out.
+    fp : FocalplaneConfig
+        This is the FocalplaneConfig object that contains info about what
+        channels are present in the focal-plane representation, and their
+        positions.
+    mask : np.ndarray
+        This is a channel mask that maps readout channel to
+        absolute-smurf-chan. Before a status frame containing the ChannelMask
+        is seen in the G3Stream, this defaults to being an identity mapping
+        which just sends the readout channel no. to itself. Once a status
+        frame with the channel mask is seen, this will be updated
+    out_queue : Queue
+        This is a queue containing outgoing G3Frames to be sent to lyrebird.
+    delay : float
+        The outgoing stream will attempt to enforce this delay between the
+        relative timestamps in the G3Frames and the real time to ensure a
+        smooth flow of data. This must be greater than the frame-aggregation
+        time of the SmurfStreamer or else lyrebird will update data in spurts.
+    avg1, avg2 : RollingAvg
+        Two Rolling Averagers which are used to calculate the rolling RMS data.
+    """
+    mask_register = 'AMCc.SmurfProcessor.ChannelMapper.Mask'
+
+    def __init__(self, agent, args):
+        self.agent = agent
+        self.log = self.agent.log
+        self._running = False
+
+        self.target_rate = args.target_rate
+        layout = args.layout.lower()
+        if layout == 'grid':
+            self.fp = FocalplaneConfig.grid(
+                args.stream_id, args.xdim, args.ydim, ygap=8, offset=args.offset
+            )
+        elif layout == 'wafer':
+            if args.det_map is not None:
+                self.fp = FocalplaneConfig.from_csv(
+                    args.stream_id, args.det_map, wafer_scale=args.wafer_scale, offset=args.offset
+                )
+            else:
+                raise ValueError("CSV file must be set using the det-map arg if "
+                                 "using wafer layout")
+
+        self.mask = np.arange(MAX_CHANS)
+        self.out_queue = queue.Queue(1000)
+        self.delay = args.delay
+
+        self.avg1, self.avg2 = None, None
+
+
+    @ocs_agent.param('target_rate', type=float)
+    def set_target_rate(self, session, params):
```

Comment:
Needs a method signature override at the top of the docstring: `set_target_rate(target_rate)`